From 384a801a17727a95fd21125a336c55ce895f879b Mon Sep 17 00:00:00 2001
From: ran shalit <rshalit@habana.ai>
Date: Mon, 31 May 2021 17:22:09 +0300
Subject: [PATCH] [SW-44018] gaudi: move e2e init to preboot

Move e2e credits initialization from u-boot to preboot, so
that even if FW stays in preboot and does not continue
into u-boot e2e credits are already initialized.

Change-Id: I5647107b52a63e0a4906b784272fe93145452aa4
Signed-off-by: ran shalit <rshalit@habana.ai>
---

diff --git a/gaudi/asic.c b/gaudi/asic.c
index a81bbc3..1e75f80 100644
--- a/gaudi/asic.c
+++ b/gaudi/asic.c
@@ -20,6 +20,28 @@
 #include <dma_if_w_s_down_ch0_regs.h>
 #include <dma_if_w_s_down_ch1_regs.h>
 #include <dma_if_w_s_regs.h>
+#include <dma_if_w_n_down_ch0_regs.h>
+#include <dma_if_w_n_down_ch1_regs.h>
+#include <dma_if_e_s_down_ch0_regs.h>
+#include <dma_if_e_s_down_ch1_regs.h>
+#include <dma_if_e_n_down_ch0_regs.h>
+#include <dma_if_e_n_down_ch1_regs.h>
+#include <nif_rtr_ctrl_0_regs.h>
+#include <nif_rtr_ctrl_1_regs.h>
+#include <nif_rtr_ctrl_2_regs.h>
+#include <nif_rtr_ctrl_3_regs.h>
+#include <nif_rtr_ctrl_4_regs.h>
+#include <nif_rtr_ctrl_5_regs.h>
+#include <nif_rtr_ctrl_6_regs.h>
+#include <nif_rtr_ctrl_7_regs.h>
+#include <sif_rtr_ctrl_0_regs.h>
+#include <sif_rtr_ctrl_1_regs.h>
+#include <sif_rtr_ctrl_2_regs.h>
+#include <sif_rtr_ctrl_3_regs.h>
+#include <sif_rtr_ctrl_4_regs.h>
+#include <sif_rtr_ctrl_5_regs.h>
+#include <sif_rtr_ctrl_6_regs.h>
+#include <sif_rtr_ctrl_7_regs.h>
 
 #include <string.h>
 #include <bitops.h>
@@ -579,3 +601,154 @@
 
 	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_HBM_CRED_EN, true);
 }
+
+void preboot_e2e_credits_init(void)
+{
+	/* The following values were determined by VLSI group. */
+	regmap_writel(247 >> 3, mmSIF_RTR_CTRL_0_E2E_HBM_WR_SIZE);
+	regmap_writel(785 >> 3, mmSIF_RTR_CTRL_0_E2E_HBM_RD_SIZE);
+	regmap_writel(49, mmSIF_RTR_CTRL_0_E2E_PCI_WR_SIZE);
+	regmap_writel(101, mmSIF_RTR_CTRL_0_E2E_PCI_RD_SIZE);
+	regmap_writel(275 >> 3, mmSIF_RTR_CTRL_1_E2E_HBM_WR_SIZE);
+	regmap_writel(614 >> 3, mmSIF_RTR_CTRL_1_E2E_HBM_RD_SIZE);
+	regmap_writel(1, mmSIF_RTR_CTRL_1_E2E_PCI_WR_SIZE);
+	regmap_writel(39, mmSIF_RTR_CTRL_1_E2E_PCI_RD_SIZE);
+	regmap_writel(1, mmSIF_RTR_CTRL_2_E2E_HBM_WR_SIZE);
+	regmap_writel(1, mmSIF_RTR_CTRL_2_E2E_HBM_RD_SIZE);
+	regmap_writel(1, mmSIF_RTR_CTRL_2_E2E_PCI_WR_SIZE);
+	regmap_writel(32, mmSIF_RTR_CTRL_2_E2E_PCI_RD_SIZE);
+	regmap_writel(176 >> 3, mmSIF_RTR_CTRL_3_E2E_HBM_WR_SIZE);
+	regmap_writel(32 >> 3, mmSIF_RTR_CTRL_3_E2E_HBM_RD_SIZE);
+	regmap_writel(19, mmSIF_RTR_CTRL_3_E2E_PCI_WR_SIZE);
+	regmap_writel(32, mmSIF_RTR_CTRL_3_E2E_PCI_RD_SIZE);
+	regmap_writel(176 >> 3, mmSIF_RTR_CTRL_4_E2E_HBM_WR_SIZE);
+	regmap_writel(32 >> 3, mmSIF_RTR_CTRL_4_E2E_HBM_RD_SIZE);
+	regmap_writel(19, mmSIF_RTR_CTRL_4_E2E_PCI_WR_SIZE);
+	regmap_writel(32, mmSIF_RTR_CTRL_4_E2E_PCI_RD_SIZE);
+	regmap_writel(1, mmSIF_RTR_CTRL_5_E2E_HBM_WR_SIZE);
+	regmap_writel(1, mmSIF_RTR_CTRL_5_E2E_HBM_RD_SIZE);
+	regmap_writel(1, mmSIF_RTR_CTRL_5_E2E_PCI_WR_SIZE);
+	regmap_writel(32, mmSIF_RTR_CTRL_5_E2E_PCI_RD_SIZE);
+	regmap_writel(275 >> 3, mmSIF_RTR_CTRL_6_E2E_HBM_WR_SIZE);
+	regmap_writel(614 >> 3, mmSIF_RTR_CTRL_6_E2E_HBM_RD_SIZE);
+	regmap_writel(1, mmSIF_RTR_CTRL_6_E2E_PCI_WR_SIZE);
+	regmap_writel(39, mmSIF_RTR_CTRL_6_E2E_PCI_RD_SIZE);
+	regmap_writel(297 >> 3, mmSIF_RTR_CTRL_7_E2E_HBM_WR_SIZE);
+	regmap_writel(908 >> 3, mmSIF_RTR_CTRL_7_E2E_HBM_RD_SIZE);
+	regmap_writel(19, mmSIF_RTR_CTRL_7_E2E_PCI_WR_SIZE);
+	regmap_writel(19, mmSIF_RTR_CTRL_7_E2E_PCI_RD_SIZE);
+	regmap_writel(318 >> 3, mmNIF_RTR_CTRL_0_E2E_HBM_WR_SIZE);
+	regmap_writel(956 >> 3, mmNIF_RTR_CTRL_0_E2E_HBM_RD_SIZE);
+	regmap_writel(79, mmNIF_RTR_CTRL_0_E2E_PCI_WR_SIZE);
+	regmap_writel(163, mmNIF_RTR_CTRL_0_E2E_PCI_RD_SIZE);
+	regmap_writel(275 >> 3, mmNIF_RTR_CTRL_1_E2E_HBM_WR_SIZE);
+	regmap_writel(614 >> 3, mmNIF_RTR_CTRL_1_E2E_HBM_RD_SIZE);
+	regmap_writel(1, mmNIF_RTR_CTRL_1_E2E_PCI_WR_SIZE);
+	regmap_writel(39, mmNIF_RTR_CTRL_1_E2E_PCI_RD_SIZE);
+	regmap_writel(1, mmNIF_RTR_CTRL_2_E2E_HBM_WR_SIZE);
+	regmap_writel(1, mmNIF_RTR_CTRL_2_E2E_HBM_RD_SIZE);
+	regmap_writel(1, mmNIF_RTR_CTRL_2_E2E_PCI_WR_SIZE);
+	regmap_writel(32, mmNIF_RTR_CTRL_2_E2E_PCI_RD_SIZE);
+	regmap_writel(176 >> 3, mmNIF_RTR_CTRL_3_E2E_HBM_WR_SIZE);
+	regmap_writel(32 >> 3, mmNIF_RTR_CTRL_3_E2E_HBM_RD_SIZE);
+	regmap_writel(19, mmNIF_RTR_CTRL_3_E2E_PCI_WR_SIZE);
+	regmap_writel(32, mmNIF_RTR_CTRL_3_E2E_PCI_RD_SIZE);
+	regmap_writel(176 >> 3, mmNIF_RTR_CTRL_4_E2E_HBM_WR_SIZE);
+	regmap_writel(32 >> 3, mmNIF_RTR_CTRL_4_E2E_HBM_RD_SIZE);
+	regmap_writel(19, mmNIF_RTR_CTRL_4_E2E_PCI_WR_SIZE);
+	regmap_writel(32, mmNIF_RTR_CTRL_4_E2E_PCI_RD_SIZE);
+	regmap_writel(1, mmNIF_RTR_CTRL_5_E2E_HBM_WR_SIZE);
+	regmap_writel(1, mmNIF_RTR_CTRL_5_E2E_HBM_RD_SIZE);
+	regmap_writel(1, mmNIF_RTR_CTRL_5_E2E_PCI_WR_SIZE);
+	regmap_writel(32, mmNIF_RTR_CTRL_5_E2E_PCI_RD_SIZE);
+	regmap_writel(275 >> 3, mmNIF_RTR_CTRL_6_E2E_HBM_WR_SIZE);
+	regmap_writel(614 >> 3, mmNIF_RTR_CTRL_6_E2E_HBM_RD_SIZE);
+	regmap_writel(1, mmNIF_RTR_CTRL_6_E2E_PCI_WR_SIZE);
+	regmap_writel(39, mmNIF_RTR_CTRL_6_E2E_PCI_RD_SIZE);
+	regmap_writel(318 >> 3, mmNIF_RTR_CTRL_7_E2E_HBM_WR_SIZE);
+	regmap_writel(956 >> 3, mmNIF_RTR_CTRL_7_E2E_HBM_RD_SIZE);
+	regmap_writel(79, mmNIF_RTR_CTRL_7_E2E_PCI_WR_SIZE);
+	regmap_writel(79, mmNIF_RTR_CTRL_7_E2E_PCI_RD_SIZE);
+	regmap_writel(344 >> 3, mmDMA_IF_E_N_DOWN_CH0_E2E_HBM_WR_SIZE);
+	regmap_writel(1000 >> 3, mmDMA_IF_E_N_DOWN_CH0_E2E_HBM_RD_SIZE);
+	regmap_writel(162, mmDMA_IF_E_N_DOWN_CH0_E2E_PCI_WR_SIZE);
+	regmap_writel(338, mmDMA_IF_E_N_DOWN_CH0_E2E_PCI_RD_SIZE);
+	regmap_writel(344 >> 3, mmDMA_IF_E_N_DOWN_CH1_E2E_HBM_WR_SIZE);
+	regmap_writel(1000 >> 3, mmDMA_IF_E_N_DOWN_CH1_E2E_HBM_RD_SIZE);
+	regmap_writel(162, mmDMA_IF_E_N_DOWN_CH1_E2E_PCI_WR_SIZE);
+	regmap_writel(338, mmDMA_IF_E_N_DOWN_CH1_E2E_PCI_RD_SIZE);
+	regmap_writel(344 >> 3, mmDMA_IF_E_S_DOWN_CH0_E2E_HBM_WR_SIZE);
+	regmap_writel(1000 >> 3, mmDMA_IF_E_S_DOWN_CH0_E2E_HBM_RD_SIZE);
+	regmap_writel(162, mmDMA_IF_E_S_DOWN_CH0_E2E_PCI_WR_SIZE);
+	regmap_writel(338, mmDMA_IF_E_S_DOWN_CH0_E2E_PCI_RD_SIZE);
+	regmap_writel(344 >> 3, mmDMA_IF_E_S_DOWN_CH1_E2E_HBM_WR_SIZE);
+	regmap_writel(1000 >> 3, mmDMA_IF_E_S_DOWN_CH1_E2E_HBM_RD_SIZE);
+	regmap_writel(162, mmDMA_IF_E_S_DOWN_CH1_E2E_PCI_WR_SIZE);
+	regmap_writel(338, mmDMA_IF_E_S_DOWN_CH1_E2E_PCI_RD_SIZE);
+	regmap_writel(344 >> 3, mmDMA_IF_W_N_DOWN_CH0_E2E_HBM_WR_SIZE);
+	regmap_writel(1000 >> 3, mmDMA_IF_W_N_DOWN_CH0_E2E_HBM_RD_SIZE);
+	regmap_writel(162, mmDMA_IF_W_N_DOWN_CH0_E2E_PCI_WR_SIZE);
+	regmap_writel(338, mmDMA_IF_W_N_DOWN_CH0_E2E_PCI_RD_SIZE);
+	regmap_writel(344 >> 3, mmDMA_IF_W_N_DOWN_CH1_E2E_HBM_WR_SIZE);
+	regmap_writel(1000 >> 3, mmDMA_IF_W_N_DOWN_CH1_E2E_HBM_RD_SIZE);
+	regmap_writel(162, mmDMA_IF_W_N_DOWN_CH1_E2E_PCI_WR_SIZE);
+	regmap_writel(338, mmDMA_IF_W_N_DOWN_CH1_E2E_PCI_RD_SIZE);
+	regmap_writel(344 >> 3, mmDMA_IF_W_S_DOWN_CH0_E2E_HBM_WR_SIZE);
+	regmap_writel(1000 >> 3, mmDMA_IF_W_S_DOWN_CH0_E2E_HBM_RD_SIZE);
+	regmap_writel(162, mmDMA_IF_W_S_DOWN_CH0_E2E_PCI_WR_SIZE);
+	regmap_writel(338, mmDMA_IF_W_S_DOWN_CH0_E2E_PCI_RD_SIZE);
+	regmap_writel(344 >> 3, mmDMA_IF_W_S_DOWN_CH1_E2E_HBM_WR_SIZE);
+	regmap_writel(1000 >> 3, mmDMA_IF_W_S_DOWN_CH1_E2E_HBM_RD_SIZE);
+	regmap_writel(162, mmDMA_IF_W_S_DOWN_CH1_E2E_PCI_WR_SIZE);
+	regmap_writel(338, mmDMA_IF_W_S_DOWN_CH1_E2E_PCI_RD_SIZE);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_0_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_0_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_1_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_1_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_2_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_2_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_3_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_3_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_4_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_4_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_5_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_5_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_6_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_6_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_7_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmSIF_RTR_CTRL_7_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_0_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_0_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_1_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_1_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_2_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_2_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_3_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_3_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_4_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_4_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_5_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_5_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_6_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_6_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_7_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmNIF_RTR_CTRL_7_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmDMA_IF_E_N_DOWN_CH0_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmDMA_IF_E_N_DOWN_CH0_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmDMA_IF_E_N_DOWN_CH1_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmDMA_IF_E_N_DOWN_CH1_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmDMA_IF_E_S_DOWN_CH0_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmDMA_IF_E_S_DOWN_CH0_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmDMA_IF_E_S_DOWN_CH1_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmDMA_IF_E_S_DOWN_CH1_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmDMA_IF_W_N_DOWN_CH0_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmDMA_IF_W_N_DOWN_CH0_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmDMA_IF_W_N_DOWN_CH1_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmDMA_IF_W_N_DOWN_CH1_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmDMA_IF_W_S_DOWN_CH0_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmDMA_IF_W_S_DOWN_CH0_E2E_PCI_EN);
+	regmap_writel(BIT(0), mmDMA_IF_W_S_DOWN_CH1_E2E_HBM_EN);
+	regmap_writel(BIT(0), mmDMA_IF_W_S_DOWN_CH1_E2E_PCI_EN);
+
+	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_E2E_CRED_EN, true);
+}
diff --git a/include/preboot.h b/include/preboot.h
index d945f8f..00bdd84 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -198,6 +198,7 @@
 
 void preboot_hbm_credits_init(void);
 void preboot_rate_limiter_init(void);
+void preboot_e2e_credits_init(void);
 #endif /* __ASSEMBLY__ */
 
 #endif /* COMMON_PREBOOT_H_ */
diff --git a/preboot/preboot.c b/preboot/preboot.c
index 16b0f2e..5e4e677 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -363,6 +363,7 @@
 __weak void preboot_process_last_rma_info(void) {}
 __weak void preboot_rate_limiter_init(void) {}
 __weak void preboot_hbm_credits_init(void) {}
+__weak void preboot_e2e_credits_init(void) {}
 
 void preboot_main(void)
 {
@@ -394,6 +395,7 @@
 	/* Get the VendorID before we lock the access to configuration space */
 	preboot_retrieve_device_id_from_pcie();
 
+	preboot_e2e_credits_init();
 	preboot_hbm_credits_init();
 	preboot_rate_limiter_init();
 
